File: problem155.py
from rational import Rational
from math import floor
from itertools import *
from itertools import chain
from tqdm import tqdm
from numpy import gcd

if __name__ == "__main__":
    limit = 18
    possible_value = [[], [(1, 1)]]
    set_poss = set({(1, 1)})

    for m in range(2, limit + 1):
        print(f"Generating every way for n={m} capacitors")
        m_value = set()

        for k in range(1, m//2 + 1):
            for (num1, denum1) in possible_value[k]:
                for (num2, denum2) in possible_value[m - k]:
                    num_prod = num1 * num2
                    denum_prod = denum1 * denum2
                    sum = num1 * denum2 + num2 * denum1
                    series_gcd = gcd(num_prod, sum)
                    parallel_gcd = gcd(sum, denum_prod)
                    series = (num_prod // series_gcd, sum // series_gcd)
                    parallel = (sum // parallel_gcd, denum_prod // parallel_gcd)
                    m_value.add(series)
                    m_value.add(parallel)

        possible_value.append(m_value)
        set_poss.update(m_value)
    print(f"There is exactly {len(set_poss)} unique ways with {limit} capacitors")

    

File: problem155.jl
using Printf

function generate_possible_value!(n::Integer, possible_value)
    push!(possible_value, [1//1])

    for m = 2:n
        @printf "Generating every way for n=%i capacitors \n" m
        m_value = []
        for k = 1:(floor(Integer, m/2))
            for C1 ∈ possible_value[k]
                for C2 ∈ possible_value[m - k]
                    parallel_capacity = C1 + C2
                    series_capacity = 1/(1/C1 + 1/C2)
                    append!(m_value, parallel_capacity)
                    append!(m_value, series_capacity)
                end
            end
        end
        unique = []
        get_unique!(m_value, unique)
        push!(possible_value, unique)
    end
end

function get_unique!(arr, out)
    flattened_array = collect(Iterators.flatten(arr))

    for val ∈ flattened_array
        if !(val ∈ out)
            push!(out, val)
        end
    end
end

function get_count(n)::Integer
    values = []
    generate_possible_value!(n, values)
    unique_values = []
    get_unique!(values, unique_values)
    return size(unique_values)[1]
end

limit = 18
possible_value = [[1//1]]
for m = 2:limit
    @printf "Generating every way for n=%i capacitors \n" m
    m_value = []
    for k = 1:(floor(Integer, m/2))
        for C1 ∈ possible_value[k]
            for C2 ∈ possible_value[m - k]
                parallel_capacity = C1 + C2
                series_capacity = 1/(1/C1 + 1/C2)
                append!(m_value, parallel_capacity)
                append!(m_value, series_capacity)
            end
        end
    end
    # unique = []
    # get_unique!(m_value, unique)
    push!(possible_value, m_value)
end

unique_values = []
get_unique!(possible_value, unique_values)
res = length(unique_values)
println("\nThere is exactly $res unique ways with $limit capacitors")